File: formula-calculator.js
/**
 * ========================================
 * FORMÜL HESAPLAYICI (Tercüman Motoru)
 * ========================================
 * 
 * Ham UDS yanıtlarını (62 10 5B 8C gibi) alır ve veritabanındaki
 * formülleri kullanarak anlamlı değerlere (54.9% gibi) dönüştürür.
 * 
 * Desteklenen formül tipleri:
 * - Basit: A, A-40, A*100/255
 * - İki byte: ((A*256)+B)/100, ((A*256)+B)
 * - Çok byte: ((A*256*256*256)+(B*256*256)+(C*256)+D)
 * - Signed (işaretli): signed((A*256)+B)
 * - Bit-encoded: Bit maskeleme
 * - Özel: Lambda, DTC kodları
 */

class FormulaCalculator {
  constructor() {
    this.debug = false;
  }

  /**
   * ⭐ ANA FONKSİYON: Ham data byte'larını formüle göre hesapla
   * 
   * @param {Array} dataBytes - UDS'den gelen ham byte'lar (HEX string array) ["8C", "A3", ...]
   * @param {string} formula - CSV'deki formül string'i
   * @param {Object} pidInfo - PID bilgileri (unit, min, max, vb.)
   * @returns {Object} { value, formatted, unit, raw }
   */
  calculate(dataBytes, formula, pidInfo = {}) {
    try {
      if (!dataBytes || dataBytes.length === 0) {
        return this.error('Veri byte\'ı yok');
      }

      if (!formula || formula.trim() === '') {
        formula = 'A'; // Varsayılan: İlk byte
      }

      // Byte'ları decimal'e çevir
      const bytes = dataBytes.map(hex => parseInt(hex, 16));
      const A = bytes[0] || 0;
      const B = bytes[1] || 0;
      const C = bytes[2] || 0;
      const D = bytes[3] || 0;

      if (this.debug) {
        console.log('🧮 Hesaplama:');
        console.log(`   Raw Bytes: ${dataBytes.join(' ')}`);
        console.log(`   Decimal: A=${A}, B=${B}, C=${C}, D=${D}`);
        console.log(`   Formül: ${formula}`);
      }

      let result;

      // Özel formül tipleri
      if (formula === 'Bit-encoded') {
        result = this.calculateBitmap(bytes);
      } else if (formula === 'DTC_CODE') {
        result = this.calculateDTC(bytes);
      } else if (formula.includes('signed')) {
        result = this.calculateSigned(formula, A, B, C, D);
      } else {
        result = this.calculateStandard(formula, A, B, C, D);
      }

      // Sonucu formatla
      const formatted = this.formatValue(result, pidInfo.unit, pidInfo.minValue, pidInfo.maxValue);

      if (this.debug) {
        console.log(`   Sonuç: ${formatted}`);
      }

      return {
        success: true,
        value: result,
        formatted: formatted,
        unit: pidInfo.unit || '',
        raw: dataBytes.join(' '),
        formula: formula
      };

    } catch (error) {
      console.error('❌ Formül hesaplama hatası:', error);
      return this.error(error.message);
    }
  }

  /**
   * Standart formül hesaplama (signed olmayan)
   */
  calculateStandard(formula, A, B, C, D) {
    // Formülü JavaScript'e çevir
    let expression = formula
      .replace(/\bA\b/g, A.toString())
      .replace(/\bB\b/g, B.toString())
      .replace(/\bC\b/g, C.toString())
      .replace(/\bD\b/g, D.toString());

    // Güvenlik: Sadece sayı, operatör ve parantez içermeli
    if (!/^[\d\s\+\-\*\/\(\)\.]+$/.test(expression)) {
      throw new Error(`Geçersiz formül: ${formula}`);
    }

    // Hesapla
    return eval(expression);
  }

  /**
   * Signed (işaretli) integer hesaplama
   * Örnek: signed((A*256)+B) -> İki byte signed 16-bit
   */
  calculateSigned(formula, A, B, C, D) {
    // "signed(...)" içindeki ifadeyi al
    const match = formula.match(/signed\(([^)]+)\)/);
    if (!match) {
      throw new Error('Signed formül formatı hatalı');
    }

    const innerFormula = match[1];
    
    // İç formülü hesapla
    const unsigned = this.calculateStandard(innerFormula, A, B, C, D);

    // Bit sayısını tespit et (formüldeki byte sayısından)
    let bits = 8;
    if (innerFormula.includes('A') && innerFormula.includes('B')) {
      bits = 16; // İki byte = 16-bit signed
    }
    if (innerFormula.includes('C') || innerFormula.includes('D')) {
      bits = 32; // 4 byte = 32-bit signed
    }

    // Unsigned'ı signed'a çevir (two's complement)
    const maxVal = Math.pow(2, bits);
    const midPoint = maxVal / 2;

    if (unsigned >= midPoint) {
      return unsigned - maxVal;
    }
    
    return unsigned;
  }

  /**
   * Bitmap hesaplama (Bit-encoded veriler)
   */
  calculateBitmap(bytes) {
    // 4-byte bitmap'i 32-bit hex string'e çevir
    const A = bytes[0] || 0;
    const B = bytes[1] || 0;
    const C = bytes[2] || 0;
    const D = bytes[3] || 0;

    const bitmap = ((A << 24) | (B << 16) | (C << 8) | D) >>> 0;
    
    return {
      hex: '0x' + bitmap.toString(16).toUpperCase().padStart(8, '0'),
      binary: bitmap.toString(2).padStart(32, '0'),
      decimal: bitmap,
      activeBits: this.getActiveBits(bitmap)
    };
  }

  /**
   * Aktif bitleri listele
   */
  getActiveBits(bitmap) {
    const active = [];
    for (let i = 0; i < 32; i++) {
      if ((bitmap >> i) & 1) {
        active.push(i + 1);
      }
    }
    return active;
  }

  /**
   * DTC (Diagnostic Trouble Code) hesaplama
   * Format: 2 byte -> P0420, C1234, vb.
   */
  calculateDTC(bytes) {
    const A = bytes[0] || 0;
    const B = bytes[1] || 0;

    // İlk 2 bit = kod tipi (P, C, B, U)
    const codeTypes = ['P', 'C', 'B', 'U'];
    const typeIndex = (A >> 6) & 0x03;
    const codeType = codeTypes[typeIndex];

    // Geri kalan 14 bit = kod numarası
    const codeNumber = ((A & 0x3F) << 8) | B;
    const codeStr = codeNumber.toString(16).toUpperCase().padStart(4, '0');

    return `${codeType}${codeStr}`;
  }

  /**
   * Değeri formatla (birim + ondalık basamak)
   */
  formatValue(value, unit, min, max) {
    // Bitmap ise özel formatlama
    if (typeof value === 'object' && value.hex) {
      return `${value.hex} (Binary: ${value.binary})`;
    }

    // DTC kodu ise direkt string
    if (typeof value === 'string' && value.match(/^[PCBU]\d{4}$/)) {
      return value;
    }

    // Sayısal değer
    if (typeof value !== 'number') {
      return String(value);
    }

    // Min/Max sınırları uygula
    if (min !== undefined && value < min) value = min;
    if (max !== undefined && value > max) value = max;

    // Ondalık basamak belirle
    let decimals = 0;
    if (Math.abs(value) < 10) decimals = 2;
    else if (Math.abs(value) < 100) decimals = 1;

    const formatted = value.toFixed(decimals);
    
    return unit ? `${formatted} ${unit}` : formatted;
  }

  /**
   * Hata döndür
   */
  error(message) {
    return {
      success: false,
      error: message,
      value: null,
      formatted: 'N/A'
    };
  }

  /**
   * Debug modunu aç/kapat
   */
  setDebug(enabled) {
    this.debug = enabled;
  }

  /**
   * ⭐ BATCH HESAPLAMA: Birden fazla PID'i aynı anda hesapla
   * 
   * @param {Array} pidDataArray - [{ bytes, formula, pidInfo }, ...]
   * @returns {Array} Hesaplama sonuçları
   */
  calculateBatch(pidDataArray) {
    return pidDataArray.map(item => {
      return this.calculate(item.bytes, item.formula, item.pidInfo);
    });
  }

  /**
   * ⭐ FORMÜL VALİDASYONU: Formül geçerli mi kontrol et
   */
  validateFormula(formula) {
    try {
      // Test değerleri ile hesapla
      const testResult = this.calculate(['FF', 'FF', 'FF', 'FF'], formula);
      return { valid: true, testResult: testResult.value };
    } catch (error) {
      return { valid: false, error: error.message };
    }
  }

  /**
   * ⭐ REVERSE CALCULATION: İstenen değer için gerekli byte'ları hesapla
   * (Örnek: 50% SOC için hangi byte gönderilmeli?)
   */
  reverseCalculate(targetValue, formula, bytePositions = 'A') {
    // Basit formüller için (A, A-40, A*100/255)
    if (formula === 'A') {
      return [Math.round(targetValue).toString(16).toUpperCase()];
    }

    if (formula === 'A-40') {
      const A = Math.round(targetValue + 40);
      return [A.toString(16).toUpperCase()];
    }

    if (formula === 'A*100/255') {
      const A = Math.round((targetValue * 255) / 100);
      return [A.toString(16).toUpperCase()];
    }

    // Daha karmaşık formüller için yaklaşık hesaplama
    // Bu kısım geliştirilmeye açık
    return null;
  }
}

// ========================================
// HAZIR FORMÜL PATTERN'LERİ (Hızlı referans)
// ========================================

const COMMON_FORMULAS = {
  // Basit tek byte
  RAW_BYTE: 'A',
  TEMPERATURE: 'A-40',              // -40°C ile 215°C arası
  PERCENTAGE: 'A*100/255',          // 0-100%
  
  // İki byte
  TWO_BYTE_UINT: '(A*256)+B',       // 16-bit unsigned
  RPM: '((A*256)+B)/4',             // Motor devri
  VOLTAGE_V: '((A*256)+B)/1000',    // Voltaj (V)
  VOLTAGE_DIV100: '((A*256)+B)/100', // Voltaj (/100)
  CURRENT_DIV10: '((A*256)+B)/10',   // Akım (/10)
  SPEED: '((A*256)+B)/100',         // Hız (km/h)
  
  // İki byte signed
  SIGNED_CURRENT: 'signed((A*256)+B)/10',  // İşaretli akım
  SIGNED_TORQUE: 'signed((A*256)+B)/10',   // İşaretli tork
  SIGNED_POWER: 'signed((A*256)+B)/10',    // İşaretli güç
  
  // Dört byte
  FOUR_BYTE_UINT: '((A*256*256*256)+(B*256*256)+(C*256)+D)',
  ENERGY_KWH: '((A*256*256*256)+(B*256*256)+(C*256)+D)/1000',
  
  // Özel
  FUEL_TRIM: '(A-128)*100/128',     // Yakıt ayarı (±100%)
  LAMBDA: '((A*256)+B)/32768',      // Lambda sensörü
  PRESSURE_BAR: '((A*256)+B)/1000', // Basınç (bar)
  
  // Bitmap
  BITMAP: 'Bit-encoded',
  DTC: 'DTC_CODE'
};

// ========================================
// KULLANIM ÖRNEKLERİ
// ========================================

/**
 * Örnek 1: Tesla SOC hesaplama
 * 
const calc = new FormulaCalculator();

// UDS yanıtı: 62 11 87 8C (62=başarı, 1187=PID, 8C=data)
const dataBytes = ['8C']; // SOC = 140 decimal = ~54.9%

const result = calc.calculate(dataBytes, 'A', {
  unit: '%',
  minValue: 0,
  maxValue: 100
});

console.log(result.formatted); // "54.90 %"
*/

/**
 * Örnek 2: Nissan Leaf batarya voltajı
 * 
const calc = new FormulaCalculator();

// UDS yanıtı: 62 01 5D 01 A3 (PID=015D, Data=01 A3)
const dataBytes = ['01', 'A3'];

const result = calc.calculate(dataBytes, '((A*256)+B)/100', {
  unit: 'V',
  minValue: 0,
  maxValue: 500
});

console.log(result.formatted); // "4.19 V" (419/100)
*/

/**
 * Örnek 3: Signed current (şarj/deşarj akımı)
 * 
const calc = new FormulaCalculator();

// Pozitif akım (şarj)
const dataBytes1 = ['00', 'C8']; // 200 decimal -> +20.0 A
const result1 = calc.calculate(dataBytes1, 'signed((A*256)+B)/10', {
  unit: 'A'
});
console.log(result1.formatted); // "20.0 A"

// Negatif akım (deşarj)
const dataBytes2 = ['FF', '38']; // -200 (two's complement) -> -20.0 A
const result2 = calc.calculate(dataBytes2, 'signed((A*256)+B)/10', {
  unit: 'A'
});
console.log(result2.formatted); // "-20.0 A"
*/

/**
 * Örnek 4: Bitmap decode (PIDs supported)
 * 
const calc = new FormulaCalculator();

const dataBytes = ['BE', '1F', 'A8', '13'];
const result = calc.calculate(dataBytes, 'Bit-encoded');

console.log(result.formatted);
// "0xBE1FA813 (Binary: 10111110000111111010100000010011)"
console.log(result.value.activeBits);
// [1, 2, 5, 9, 10, 11, 12, 14, 15, 16, 17, 18, 19, 22, 24, 26, 28, 32]
*/

/**
 * Örnek 5: DTC kod okuma
 * 
const calc = new FormulaCalculator();

const dataBytes = ['01', '14']; // P0114 (IAT circuit intermittent)
const result = calc.calculate(dataBytes, 'DTC_CODE');

console.log(result.formatted); // "P0114"
*/

/**
 * Örnek 6: Debug modu ile detaylı log
 * 
const calc = new FormulaCalculator();
calc.setDebug(true);

const result = calc.calculate(['8C'], 'A*100/255', { unit: '%' });
// Console'da detaylı hesaplama adımları gösterilir
*/

/**
 * Örnek 7: Batch hesaplama (çoklu PID)
 * 
const calc = new FormulaCalculator();

const batchData = [
  { bytes: ['8C'], formula: 'A', pidInfo: { unit: '%' } },
  { bytes: ['01', 'A3'], formula: '((A*256)+B)/100', pidInfo: { unit: 'V' } },
  { bytes: ['FF', '38'], formula: 'signed((A*256)+B)/10', pidInfo: { unit: 'A' } }
];

const results = calc.calculateBatch(batchData);
results.forEach(r => console.log(r.formatted));
// "54.90 %"
// "4.19 V"
// "-20.0 A"
*/

// Export
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { FormulaCalculator, COMMON_FORMULAS };
}

File: formula_calculator.py
"""
Formül Hesaplayıcı Modülü
UDS cevaplarını formüllere göre işleyip anlamlı veriye çevirir
"""

import re
import math
from typing import Optional, Union, List, Dict, Any
from pid_router import PIDEntry


class FormulaCalculator:
    """UDS cevaplarını formüllere göre hesaplayan sınıf"""
    
    def __init__(self):
        """Hesaplayıcıyı başlat"""
        self.byte_cache: Dict[str, int] = {}
    
    def calculate(self, pid_entry: PIDEntry, response_bytes: List[int]) -> Optional[Union[float, int, str]]:
        """
        UDS cevabını formüle göre hesapla
        
        Args:
            pid_entry: PIDEntry nesnesi
            response_bytes: UDS'den gelen ham byte listesi (örn: [0x62, 0x10, 0x5B, 0x8C])
        
        Returns:
            Hesaplanmış değer (float, int veya string)
        
        Örnek:
            response = [0x62, 0x10, 0x5B, 0x8C]
            # 0x62 = positive response (atla)
            # 0x10 0x5B = echo (atla)
            # 0x8C = gerçek veri (140 decimal)
            # Formül: A (140 * 100 / 255 = 54.9%)
        """
        if not response_bytes or len(response_bytes) < 3:
            return None
        
        try:
            # UDS positive response kontrolü
            # Format: [0x62, service_echo, pid_byte1, pid_byte2, data...]
            if response_bytes[0] == 0x62:
                # Positive response, echo'yu atla
                data_start = 3  # 0x62 + service + pid
            elif response_bytes[0] == 0x7F:
                # Negative response
                return None
            else:
                # Direkt veri
                data_start = 0
            
            # Byte pozisyonlarını parse et
            byte_positions = self._parse_byte_positions(pid_entry.byte_positions)
            
            # Byte değerlerini al
            byte_values = {}
            for pos, idx in byte_positions.items():
                if data_start + idx < len(response_bytes):
                    byte_values[pos] = response_bytes[data_start + idx]
                else:
                    byte_values[pos] = 0
            
            # Formülü hesapla
            result = self._evaluate_formula(pid_entry.equation, byte_values)
            
            # Min/Max kontrolü (opsiyonel)
            if pid_entry.min_val and pid_entry.max_val:
                try:
                    min_val = float(pid_entry.min_val)
                    max_val = float(pid_entry.max_val)
                    if isinstance(result, (int, float)):
                        result = max(min_val, min(max_val, result))
                except ValueError:
                    pass
            
            return result
            
        except Exception as e:
            print(f"Hesaplama hatası ({pid_entry.name}): {e}")
            return None
    
    def _parse_byte_positions(self, byte_positions_str: str) -> Dict[str, int]:
        """
        Byte pozisyonlarını parse et
        
        Args:
            byte_positions_str: "A-B" veya "A-B-C-D" formatında string
        
        Returns:
            {'A': 0, 'B': 1, ...} formatında dict
        """
        positions = {}
        parts = byte_positions_str.upper().replace(' ', '').split('-')
        
        for i, part in enumerate(parts):
            if part:
                positions[part] = i
        
        return positions
    
    def _evaluate_formula(self, formula: str, byte_values: Dict[str, int]) -> Union[float, int, str]:
        """
        Formülü değerlendir
        
        Args:
            formula: Formül string'i (örn: "A*100/255", "(A*256+B)/100")
            byte_values: Byte değerleri dict'i
        
        Returns:
            Hesaplanmış değer
        """
        if not formula or formula.strip() == '':
            # Formül yoksa ilk byte'ı döndür
            return byte_values.get('A', 0)
        
        # Özel durumlar
        formula_upper = formula.upper().strip()
        
        if formula_upper == 'ASCII':
            # ASCII string
            return self._decode_ascii(byte_values)
        elif formula_upper == 'BIT ENCODED' or formula_upper == 'BIT':
            # Bit encoded
            return byte_values.get('A', 0)
        elif formula_upper == 'RAW DATA' or formula_upper == 'HEX':
            # Ham hex data
            return self._format_hex(byte_values)
        elif formula_upper == 'DATE':
            # Tarih formatı
            return self._decode_date(byte_values)
        elif 'RANDOM' in formula_upper or 'CALCULATED' in formula_upper:
            # Rastgele veya hesaplanmış değer
            return byte_values.get('A', 0)
        
        # Matematiksel formül
        try:
            # Formüldeki değişkenleri byte değerleriyle değiştir
            formula_processed = formula
            
            # Byte değerlerini yerleştir (A, B, C, D)
            for byte_name in ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']:
                value = byte_values.get(byte_name, 0)
                # Regex ile güvenli değiştirme
                pattern = r'\b' + byte_name + r'\b'
                formula_processed = re.sub(pattern, str(value), formula_processed)
            
            # Güvenli eval (sadece matematiksel işlemler)
            # Tehlikeli fonksiyonları kaldır
            allowed_names = {
                "__builtins__": {},
                "abs": abs,
                "max": max,
                "min": min,
                "round": round,
                "int": int,
                "float": float,
                "math": math,
                "sqrt": math.sqrt,
                "sin": math.sin,
                "cos": math.cos,
                "tan": math.tan,
                "pi": math.pi,
                "e": math.e
            }
            
            result = eval(formula_processed, allowed_names)
            
            # Sonucu uygun formata çevir
            if isinstance(result, float):
                # Çok küçük değerleri 0 yap
                if abs(result) < 1e-10:
                    result = 0.0
                # Gereksiz ondalık basamakları temizle
                if result.is_integer():
                    return int(result)
            
            return result
            
        except Exception as e:
            print(f"Formül hesaplama hatası ({formula}): {e}")
            # Hata durumunda ilk byte'ı döndür
            return byte_values.get('A', 0)
    
    def _decode_ascii(self, byte_values: Dict[str, int]) -> str:
        """ASCII byte'ları string'e çevir"""
        chars = []
        for byte_name in ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']:
            value = byte_values.get(byte_name, 0)
            if 32 <= value <= 126:  # Yazdırılabilir ASCII
                chars.append(chr(value))
            elif value == 0:
                break
        return ''.join(chars).strip()
    
    def _format_hex(self, byte_values: Dict[str, int]) -> str:
        """Byte'ları hex string'e çevir"""
        hex_parts = []
        for byte_name in ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']:
            value = byte_values.get(byte_name)
            if value is not None:
                hex_parts.append(f"{value:02X}")
        return ' '.join(hex_parts)
    
    def _decode_date(self, byte_values: Dict[str, int]) -> str:
        """Tarih byte'larını decode et"""
        # Basit tarih formatı (byte formatına göre değişebilir)
        year = byte_values.get('A', 0) + 2000
        month = byte_values.get('B', 0)
        day = byte_values.get('C', 0)
        return f"{year:04d}-{month:02d}-{day:02d}"
    
    def format_value(self, value: Union[float, int, str], units: str, pid_entry: PIDEntry) -> str:
        """
        Değeri formatla ve birimle birlikte göster
        
        Args:
            value: Hesaplanmış değer
            units: Birim (örn: "°C", "%", "V")
            pid_entry: PIDEntry nesnesi
        
        Returns:
            Formatlanmış string (örn: "54.9%", "25.5°C")
        """
        if value is None:
            return "N/A"
        
        if isinstance(value, str):
            return value
        
        # Birim formatlaması
        if units == '%':
            return f"{value:.1f}%"
        elif units == '°C' or units == '°F':
            return f"{value:.1f}{units}"
        elif units == 'V':
            return f"{value:.2f}V"
        elif units == 'A':
            return f"{value:.2f}A"
        elif units == 'kW':
            return f"{value:.2f}kW"
        elif units == 'kWh':
            return f"{value:.2f}kWh"
        elif units == 'rpm':
            if value.is_integer():
                return f"{int(value)}rpm"
            return f"{value:.1f}rpm"
        elif units == 'km/h':
            return f"{value:.1f}km/h"
        elif units == 'kPa' or units == 'bar':
            return f"{value:.2f}{units}"
        elif units == 'Nm':
            return f"{int(value)}Nm"
        elif units == 'Count':
            return f"{int(value)}"
        elif units == 'Status':
            return f"{int(value)}"
        else:
            # Genel format
            if isinstance(value, float):
                if value.is_integer():
                    return f"{int(value)}{units}"
                return f"{value:.2f}{units}"
            return f"{value}{units}"
    
    def calculate_with_formatting(self, pid_entry: PIDEntry, response_bytes: List[int]) -> Optional[str]:
        """
        Hesapla ve formatla (tek adımda)
        
        Args:
            pid_entry: PIDEntry nesnesi
            response_bytes: UDS cevap byte'ları
        
        Returns:
            Formatlanmış string (örn: "54.9%")
        """
        value = self.calculate(pid_entry, response_bytes)
        if value is None:
            return None
        
        return self.format_value(value, pid_entry.units, pid_entry)
